addons/binding/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/handler/SleepIQCloudHandler.java

```diff
@@ -0,0 +1,322 @@
+     *
+     * @throws LoginException if there is an error while authenticating to the service
+     */
+    private synchronized void createCloudConnection() throws LoginException {
+
+        logger.debug("Reading SleepIQ cloud binding configuration");
+        SleepIQCloudConfiguration bindingConfig = getConfigAs(SleepIQCloudConfiguration.class);
+
+        logger.debug("Creating SleepIQ client");
+        Configuration cloudConfig = new Configuration().withUsername(bindingConfig.username)
+                .withPassword(bindingConfig.password).withLogging(bindingConfig.logging);
+        cloud = SleepIQ.create(cloudConfig);
+
+        logger.info("Authenticating to SleepIQ cloud service");
+        cloud.login();
+
+        logger.info("Authentication successful");
+    }
+
+    @Override
+    public synchronized void dispose() {
+
+        logger.debug("Disposing SleepIQ cloud handler");
+
+        if (pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+
+        cloud = null;
+    }
+
+    /**
+     * Get the cloud service API.
+     *
+     * @return the service object (may be <code>null</code>)
+     */
+    private synchronized SleepIQ getCloud() {
+        return cloud;
+    }
+
+    /**
+     * Start or stop a background polling job to look for bed status updates based on the whether or not there are any
+     * listeners to notify.
+     */
+    private synchronized void updateListenerManagement() {
+
+        if (cloud == null) {
+            return;
+        }
+
+        if (!bedStatusListeners.isEmpty() && (pollingJob == null || pollingJob.isCancelled())) {
+            int pollingInterval = getConfigAs(SleepIQCloudConfiguration.class).pollingInterval;
+            pollingJob = scheduler.scheduleAtFixedRate(pollingRunnable, pollingInterval, pollingInterval,
+                    TimeUnit.SECONDS);
+        } else if (bedStatusListeners.isEmpty() && pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update all registered listeners.
+     */
+    private void publishBedStatusUpdates() {
+        publishBedStatusUpdates(null);
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update the given listener. If the listener is <code>null</code>, all
+     * registered listeners will be updated.
+     *
+     * @param listener the listener to update (may be <code>null</code>)
+     */
+    private void publishBedStatusUpdates(final BedStatusListener listener) {
+
+        SleepIQ cloud = getCloud();
+        if (cloud == null) {
+            return;
+        }
+
+        FamilyStatus status = cloud.getFamilyStatus();
+        for (BedStatus bedStatus : status.getBeds()) {
+
+            if (listener != null) {
+                listener.onBedStateChanged(cloud, bedStatus);
```

Comment:
There is only one type of listener - BedStatusListener. My goal is to avoid retrieving the status (HTTP call) more than once. I have two callers for this method (one with a specific listener to update and the other wanting to broadcast the update to all registered listeners). The former passes in an instance of BedStatusListener and the latter passes null triggering the broadcast to all listeners.

An alternative approach would be to retrieve the status in each of the callers and pass that to this method. In that way, I could have made this method accept a listener and a status. However, at that point this method isn't really doing anything and I don't need it because I have basically duplicated the logic within each caller.